src/Client/HedgedConnectionsFactory.cpp

```diff
@@ -0,0 +1,360 @@
+        if (remove_from_epoll)
+            removeReplicaFromEpoll(index);
+
+        if (replica.connection_establisher.getResult().is_usable)
+        {
+            ++usable_count;
+            if (replica.connection_establisher.getResult().is_up_to_date)
+                replica.is_ready = true;
+
+            return;
+        }
+
+        /// If replica is not usable, we need to save fail message.
+        if (!replica.connection_establisher.getFailMessage().empty())
+            fail_messages += replica.connection_establisher.getFailMessage() + "\n";
+    }
+    else if (replica.connection_establisher.isFailed())
+        processFailedConnection(index, remove_from_epoll);
+}
+
+void HedgedConnectionsFactory::processFailedConnection(int index, bool remove_from_epoll)
+{
+    ConnectionEstablisher & connection_establisher = replicas[index].connection_establisher;
+
+    if (remove_from_epoll)
+        removeReplicaFromEpoll(index);
+
+    if (!connection_establisher.getFailMessage().empty())
+        fail_messages += connection_establisher.getFailMessage() + "\n";
+
+    ShuffledPool & shuffled_pool = shuffled_pools[index];
+    LOG_WARNING(
+        log, "Connection failed at try №{}, reason: {}", (shuffled_pool.error_count + 1), connection_establisher.getFailMessage());
+    ProfileEvents::increment(ProfileEvents::DistributedConnectionFailTry);
+
+    shuffled_pool.error_count = std::min(pool->getMaxErrorCup(), shuffled_pool.error_count + 1);
+
+    if (shuffled_pool.error_count >= max_tries)
+    {
+        ++failed_pools_count;
+        ProfileEvents::increment(ProfileEvents::DistributedConnectionFailAtAll);
+    }
+
+    --replicas_in_process_count;
+}
+
+HedgedConnectionsFactory::State HedgedConnectionsFactory::processEpollEvents(bool blocking, Connection *& connection_out)
+{
+    int event_fd;
+    while (!epoll.empty())
+    {
+        event_fd = getReadyFileDescriptor(blocking);
+
+        if (event_fd == -1)
+            return State::NOT_READY;
+
+        if (fd_to_replica_index.contains(event_fd))
+        {
+            int index = fd_to_replica_index[event_fd];
+            processConnectionEstablisherEvent(index, connection_out);
+
+            if (replicas[index].is_ready)
+                return State::READY;
+            if (replicas[index].connection_establisher.isInProcess())
+                continue;
+        }
+        else if (timeout_fd_to_replica_index.contains(event_fd))
+            replicas[timeout_fd_to_replica_index[event_fd]].change_replica_timeout.reset();
+        else
+            throw Exception("Unknown event from epoll", ErrorCodes::LOGICAL_ERROR);
+
+        /// We reach this point only if we need to start new connection
+        /// (Special timeout expired or one of the previous connections failed).
+        int index = startEstablishingNewConnection(connection_out);
+
+        /// Return only if replica is ready.
+        if (index != -1 && replicas[index].is_ready)
+            return State::READY;
```

Comment:
Maybe move this code under `else if (timeout_fd_to_replica_index.contains(event_fd))` condition?
I think it will be more readable.

Reply:
But we start new connection not only when change_replica_timeout expires, but also when one of the connection fails.

Reply:
Indeed.
It was a little bit tricky.